drivers/comparator/comparator_it51xxx_vcmp.c

```diff
@@ -0,0 +1,451 @@
+						 REG_VCMP_VCMP2CTL};
+
+/* 0x21, 0x29, 0x2D: Voltage Comparator n Status and Control Register  (VCMPnSCTL) (n=0 to 2) */
+#define REG_VCMP_VCMP0SCTL 0x21
+#define REG_VCMP_VCMP1SCTL 0x29
+#define REG_VCMP_VCMP2SCTL 0x2D
+#define VCMP_CMPXRTIS      BIT(6)
+const uint8_t vcmp_status_ctrl_reg[VCMP_CHANNEL_CNT] = {REG_VCMP_VCMP0SCTL, REG_VCMP_VCMP1SCTL,
+							REG_VCMP_VCMP2SCTL};
+
+/* 0x22, 0x2A, 0x2E: Voltage Comparator 0~2 MSB Threshold Data Buffer (10-bit resolution) */
+#define REG_VCMP_CH_THRDATM 0x02
+
+/* 0x23, 0x2B, 0x2F: Voltage Comparator 0~2 LSB Threshold Data Buffer (10-bit resolution) */
+#define REG_VCMP_CH_THRDATL 0x03
+
+/* 0x33: Voltage Comparator Scan Period 2 (VCMPSCP2) */
+#define REG_VCMP_VCMPSCP2 0x33
+#define SCAN_PERIOD_MASK  GENMASK(7, 4)
+
+/* Device config */
+struct vcmp_it51xxx_config {
+	/* Voltage comparator channel base address */
+	uintptr_t base_ch;
+	/* Voltage comparator base address */
+	uintptr_t reg_base;
+	/* Voltage comparator module irq */
+	int irq;
+	/* Voltage comparator channel */
+	int vcmp_ch;
+	/* Comparator 0/1/2 Scan Period */
+	uint8_t scan_period;
+	/* Threshold assert value in mv */
+	int32_t threshold_mv;
+	/* Pointer of ADC device that will be performing measurement */
+	const struct device *adc;
+	/* Channel identifier */
+	uint8_t channel_id;
+};
+
+/* Driver data */
+struct vcmp_it51xxx_data {
+	/* ADC channel config */
+	struct adc_channel_cfg adc_ch_cfg;
+	/* Work queue to be notified when threshold assertion happens */
+	struct k_work work;
+	comparator_callback_t user_cb;
+	void *user_cb_data;
+	uint8_t interrupt_mask;
+	/* Pointer of voltage comparator device */
+	const struct device *vcmp;
+};
+/*
+ * Because all the three voltage comparators use the same interrupt number, the
+ * 'irq_connect_dynamic' call in the driver init function sets the passed-in dev pointer in isr
+ * function to the first instantiated voltage comparator.
+ *
+ * For example, when setting vcmp2 and vcmp3 to 'status = "okay";' in dts file, each time an
+ * interrupt is triggered,the passed-in 'const struct device *dev' argument in isr function always
+ * points to vcmp2 device.
+ *
+ * To access the 'struct vcmp_it51xxx_data' of respective voltage comparator instance, we
+ * use a array to store their address.
+ */
+static struct vcmp_it51xxx_data *vcmp_data[VCMP_CHANNEL_CNT];
+
+/* Voltage comparator work queue address */
+static uint32_t vcmp_work_addr[VCMP_CHANNEL_CNT];
+#ifdef CONFIG_VCMP_IT51XXX_WORKQUEUE
+struct k_work_q vcmp_it51xxx_work_q;
+
+static K_KERNEL_STACK_DEFINE(vcmp_it51xxx_work_q_stack, CONFIG_VCMP_IT51XXX_WORKQUEUE_STACK_SIZE);
+
+static int vcmp_it51xxx_init_work_q(void)
+{
+	struct k_work_queue_config cfg = {
+		.name = "vcmp_work",
+		.no_yield = false,
+	};
+
+	k_work_queue_start(&vcmp_it51xxx_work_q, vcmp_it51xxx_work_q_stack,
+			   K_KERNEL_STACK_SIZEOF(vcmp_it51xxx_work_q_stack),
+			   CONFIG_VCMP_IT51XXX_WORKQUEUE_PRIORITY, &cfg);
+
+	return 0;
+}
```

Comment:
Work queues are the most expensive primitive we have in zephyr, to the point where we have to share them to justify them (see system workqueue), if a driver needs one, use the system work queue, however, this driver does not need it, all execution can be done from ISR context.